File: node/sorting/bubbleSort.js
// creates an ascending sort via bubble sort
// big O notation for time complexity is O(N^2) or quadratic

function bubbleSort(arr) {
  let stepCount = 0;
  let lastIndexToCheck = arr.length - 2;

  while (true) {
    let switchCount = 0;
    for (let i = 0; i <= lastIndexToCheck; i++) {
      stepCount++;
      let nextIndex = i + 1;
      if (arr[i] > arr[nextIndex]) {
        // reverse the values of the two indices
        let tempVal = arr[i];
        arr[i] = arr[nextIndex];
        arr[nextIndex] = tempVal;
        switchCount++;
      }
    }
    lastIndexToCheck--;
    if (switchCount === 0 || lastIndexToCheck < 0) {
      break;
    }
  }
  console.log('sorted array:', arr);
  console.log('number of steps:', stepCount);
}

let arrie1 = [5, 4, 3, 2, 1];
bubbleSort(arrie1);

let arrie2 = [1, 2, 3, 4, 5];
bubbleSort(arrie2);

File: python/sorting/bubble_sort.py
"""
bubble sort algorithm

time complexity = O(N^2) or quadratic
"""

def bubble_sort(arr):
    # initially the second to last index,
    # keeps from attempting switch with index past the end of the array/list
    last_index_to_check = len(arr) - 2

    while last_index_to_check >= 0 or switch_counter != 0:
        switch_counter = 0

        for current_i, _ in enumerate(arr):
            if current_i > last_index_to_check:
                break
            else:
                next_i = current_i + 1
                if arr[next_i] < arr[current_i]:
                    switch_counter += 1
                    arr[current_i], arr[next_i] = arr[next_i], arr[current_i]

        last_index_to_check -= 1

    return arr
